src/App.tsx

```diff
@@ -1,26 +1,309 @@
 /* eslint-disable max-len */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useEffect, useState, useRef } from 'react';
 import { UserWarning } from './UserWarning';
+import { USER_ID, getTodos, addTodo, deleteTodo } from './api/todos';
+import { Todo } from './types/Todo';
 
-const USER_ID = 0;
+enum Filter {
+  All = 'all',
+  Active = 'active',
+  Completed = 'completed',
+}
+
+enum ErrorMessage {
+  Load = 'Unable to load todos',
+  Add = 'Unable to add a todo',
+  Delete = 'Unable to delete a todo',
+  EmptyTitle = 'Title should not be empty',
+}
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [tempTodoId, setTempTodoId] = useState<number | null>(null);
+  const [title, setTitle] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [isLoading, setIsLoading] = useState(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (errorMessage) {
+      const timer = setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+
+      return () => clearTimeout(timer);
+    }
+
+    return undefined;
+  }, [errorMessage]);
+
+  useEffect(() => {
+    if (!USER_ID) {
+      return;
+    }
+
+    setIsLoading(true);
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage(ErrorMessage.Load))
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  useEffect(() => {
+    inputRef.current?.focus();
+  }, [todos]);
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+    const trimmed = title.trim();
+
+    if (!trimmed) {
+      setErrorMessage(ErrorMessage.EmptyTitle);
+
+      return;
+    }
+
+    const id = Date.now();
+    const newTodo: Todo = {
+      id,
+      userId: USER_ID,
+      title: trimmed,
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+    setTempTodoId(id);
+    setIsLoading(true);
+
+    addTodo(trimmed)
+      .then(createdTodo => {
+        setTodos(prev => [...prev, createdTodo]);
+        setTitle('');
+      })
+      .catch(() => setErrorMessage(ErrorMessage.Add))
+      .finally(() => {
+        setTimeout(() => setTempTodo(null), 700);
+      });
+  };
+
+  const handleDelete = (todoId: number) => {
+    setIsLoading(true);
+    deleteTodo(todoId)
+      .then(() => setTodos(prev => prev.filter(t => t.id !== todoId)))
+      .catch(() => setErrorMessage(ErrorMessage.Delete))
+      .finally(() => setIsLoading(false));
+  };
+
+  const handleClearCompleted = () => {
+    const completed = todos.filter(t => t.completed);
+
+    if (completed.length === 0) {
+      return;
+    }
+
+    Promise.allSettled(completed.map(t => deleteTodo(t.id))).then(results => {
```

Comment:
```suggestion
    Promise.allSettled(completed.map(t => deleteTodo(t.id))).then(results => {
```
use handleDelete function here.